### PR TITLE
[CDAP-11618][CDAP-11621] DataPrep Kafka Connection

### PR DESCRIPTION
**Pull request by elfenheart**
JIRAs:
- https://issues.cask.co/browse/CDAP-11621
- https://issues.cask.co/browse/CDAP-11618

### There are still a few things not working:
~1. Currently the connectionId is being hardcoded as `testkafka` for when you add to pipeline. This is because the backend app is not returning the `connectionid` as part of the workspace properties.~
2. I have not added the ability for one time copy yet. There are properties missing for batch pipeline that is not macro enabled.
3. Existence of Kafka Plugins is still an open question....

#### Side note:
We should be able to refactor the browser to be a generic browser, especially when it only has 1 column. All the functionality is the same except the API to prep the content before going to workspace

### CODE REVIEW

File: cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/DataPrepBrowserStore/ActionCreator.js [hunk 1]
```diff
@@ -52,15 +69,59 @@ const setDatabaseAsActiveBrowser = (payload) => {
...
-      let properties = objectQuery(res, 'values', 0, 'properties');
+      let info = objectQuery(res, 'values', 0);
 
-      setDatabaseProperties({
-        properties,
-        connectionId: params.connectionId
+      MyDataPrepApi.listTables(params)
+        .subscribe((tables) => {
+          setDatabaseProperties({
+            info,
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/DataPrepBrowserStore/ActionCreator.js:77]: One thing that might happen is we might have `info` but listTables call might fail. So we might end up not showing anything. Can we do setDatabaseProperties outside for info and then for tables inside listTables?

File: cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/KafkaBrowser/index.js [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+      namespace,
+      connectionId: this.state.connectionId,
+      topic,
+      lines: 100
+    };
+
+    MyDataPrepApi.readTopic(params)
+      .subscribe(
+        (res) => {
+          console.log('test', res);
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/KafkaBrowser/index.js]: nit: do we need this?

File: cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/KafkaBrowser/KafkaBrowser.scss [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      font-size: 18px;
+      font-weight: 500;
+      margin-left: auto;
+      margin-right: auto;
+      margin-top: 25vh;
+      max-width: 80%;
+      word-wrap: break-word;
+
+      hr {
+        color: #999999;
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/KafkaBrowser/KafkaBrowser.scss]: Might as well extract this out too

File: cdap-ui/app/cdap/components/DataPrep/TopPanel/CreateDatasetBtn/index.js [hunk 4]
```diff
@@ -320,6 +337,8 @@ export default class CreateDatasetBtn extends Component {
             namespace,
             appId: pipelineName
           };
+
+          console.log('config', pipelineconfig);
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/DataPrep/TopPanel/CreateDatasetBtn/index.js]: nit: Remove?

File: cdap-ui/app/cdap/components/DataPrepConnections/KafkaConnection/index.js [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+          loading: false
+        });
+      });
+  }
+
+  parseBrokers(brokers) {
+    let brokersList = [];
+
+    brokers.split(',').forEach((broker) => {
+      let split = broker.trim().split(':');
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/DataPrepConnections/KafkaConnection/index.js:99]: You don't have to do this but since this seems to be pattern that is used in a few places, you can shorten this to something like this,

```
let [host='', port=''] = broker.trim.split(':');
let obj = {host, port, uniqueId: shortid.generate()};
...
```

File: cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/DataPrepBrowserStore/ActionCreator.js [hunk 6]
```diff
@@ -52,15 +69,59 @@ const setDatabaseAsActiveBrowser = (payload) => {
...
+  };
+
+  MyDataPrepApi.getConnection(params)
+    .subscribe((res) => {
+      let info = objectQuery(res, 'values', 0);
+
+      MyDataPrepApi.listTopics({namespace}, info)
+        .subscribe((topics) => {
+          setKafkaProperties({
+            info,
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/DataPrepBrowserStore/ActionCreator.js:112]: Similar to `fetchTables` here as well.

**Review by ajainarayanan — COMMENTED** [review]
Just couple of minor comments. Otherwise code LGTM 👍

**Comment by bdmogal** [comment]
When Kafka is configured as a connection, but is down, data prep keeps spinning, until you finally get this message - 
image

**Comment by bdmogal** [comment]
When the user clicks Test Connection or Add Connection in the Kafka Modal, it should first reset the contents of error/success message div

**Comment by bdmogal** [comment]
@ajainarayanan placeholder for search box is wrong. It should be ``Search topic``

**Comment by bdmogal** [comment]
@ajainarayanan sometimes when I click on a topic, it doesn't open up in data prep. When I keep clicking on topics, eventually the first topic I clicked on opens. Then after a while the second opens. Then the third after some more time...

File: cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/DataPrepBrowserStore/ActionCreator.js (hunk 1 above)
Comment (ajainarayanan) [reply cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/DataPrepBrowserStore/ActionCreator.js:77]: Not modifying this as this will prevent repeated rendering of component subscribing to the store.

**Comment by bdmogal** [comment]
@ajainarayanan "Test Connection" should also be disabled when name is not provided

**Comment by bdmogal** [comment]
Success connected to Kafka -> Successfully connected to Kafka

**Comment by ajainarayanan** [comment]
@bdmogal The success message is coming from backend. UI displays it as is.

**Comment by bdmogal** [comment]
@ajainarayanan when you get an error, Preparing this task should show a cross
image

**Comment by bdmogal** [comment]
@ajainarayanan let's add key bindings to Ingest Data modal.

**Comment by bdmogal** [comment]
@ajainarayanan looks like ingest data is failing because kafka source has wrong/incomplete schema

image

**Comment by bdmogal** [comment]
1. message -> body
2. topic name should not be used as reference name
3. change format to text, and add a set-charset by default.

**Comment by bdmogal** [comment]
There is a bug with kafka realtime source which causes realtime pipelines to fail. Other than that, this works after https://github.com/hydrator/wrangler/pull/138

**Review by bdmogal — APPROVED** [review]
